### PR TITLE
lib/os: Add sys_multi_heap utility

### PR DESCRIPTION
**Pull request by andyross**
This is a simple wrapper allowing multiple sys_heap regions to be unified under a single allocation API.  Sometimes apps need the ability to share multiple discontiguous regions in a single "heap", or to have memory of different "types" be allocated heuristically based on usage (e.g. cacheability, latency, power...).  This allows a user-specified function to select the underlying memory to use for each application.

### CODE REVIEW

**Review by npitre — COMMENTED** [review]
Looks fine, except for one detail:
You should pass the pointer for the corresponding struct sys_multi_heap
down to the choice function as well. On many architectures that pointer
will already be in the right register upon entering
`sys_multi_heap_aligned_alloc()` anyway so it's actually more expensive not to
pass it.

**Comment by andyross** [comment]
That's... an excellent point.  And I swear when I specified this originally (for an internal consumer) it was doing that.  I'm wondering where it got dropped...

**Comment by carlocaione** [comment]
LGTM and this will come in handy for a work I was planning to do.

I'm also very interested in the note:
```
In general a multiheap is likely to be instantiated
semi-statically from system configuration (for example, via
linker-provided bounds on available memory in different regions, or
from devicetree definitions of hardware-provided addressible
memory, etc...).  The general expectation is that a soc- or
board-level platform device will be initialized at system boot from
these upstream configuration sources and not that an application
will assemble a multi-heap on its own.
 ```

What about using some sub-system or OS service for that? For example leveraging the `reserved-memory` work, we could construct the multi-heap cycling through all the memory regions marked with some DT property (something like `multi-heap`).

Some other property (for example `cacheable / non-cacheable`) of the region would be used by the `choice` function to return the correct heap.

**Review by peter-mitsis — COMMENTED** [review]
Am I correct in thinking that the routines sys_multi_heap_add_heap() and sys_multi_heap_free() are not intended to be mutli-thread safe?

**Comment by andyross** [comment]
Right, this is a wrapper for sys_heap, which is the unsynchronized heap utility, not k_heap which provides synchronization (and also blocking/retry behavior, which is sort of unique to Zephyr).   It's not clear to me if wrapping it the same way we do for k_heap has much value, given that it's always going to be deployed as part of an application framework.

**Comment by andyross** [comment]
Hrm... remaining MAINTAINERS failure seems to be a false positive?

**Comment by nashif** [comment]
> Hrm... remaining MAINTAINERS failure seems to be a false positive?

you need to rebase.

File: lib/os/multi_heap.c [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+#include <sys/sys_heap.h>
+#include <sys/multi_heap.h>
+
+void sys_multi_heap_init(struct sys_multi_heap *heap, sys_multi_heap_fn_t choice_fn)
+{
+	heap->nheaps = 0;
+	heap->choice = choice_fn;
+}
+
+void sys_multi_heap_add_heap(struct sys_multi_heap *mheap, struct sys_heap *heap)
```
Comment (nashif) [inline lib/os/multi_heap.c:15]: maybe drop the last _heap in the name?

**Comment by andyross** [comment]
(rebase, turned I'd somehow put this on a two-week-stale tree, oops)

> What about using some sub-system or OS service for that?

Yeah, this is sort of a hole we have right now.  There are like nine different ways we use the linker to calculate bounds for heap-like memory in different contexts, and in general only on a handful of platforms.  It would be really nice to have a framework for this, even if it's a little out of scope here.

But I guess I'll note that the specific hardware being targetted by this work (the intel_adsp family of DSPs) is so complicated that I doubt a generic framework is every going to help (Four different memory types!  One of which has translation capability and bank-level power control, one is an outrageously slow window into system dram, one is low power, even one that is CPU-local.  And all of those can be selectively accessed via the L1 cache or via a bypass mapping!).   So there's a limit to how much something like that can help.

**Review by dcpleung — APPROVED** [review]

**Review by carlocaione — APPROVED** [review]

**Review by npitre — APPROVED** [review]

**Review by nashif — CHANGES_REQUESTED** [review]
Please document this new utility in  doc/reference/kernel/memory/heap.rst

**Comment by andyross** [comment]
Doc section added

**Review by peter-mitsis — CHANGES_REQUESTED** [review]
For the commit ... doc: lib/os: Add overview docs for sys_multi_heap

Line 111: Typo. initilized -> initialized

Line 124. Punctuation. "itself, instead" -> "itself. Instead"

**Comment by nashif** [comment]
> For the commit ... doc: lib/os: Add overview docs for sys_multi_heap
> 
> Line 111: Typo. initilized -> initialized
> 
> Line 124. Punctuation. "itself, instead" -> "itself. Instead"

you can such comments inline :)

**Review by nashif — APPROVED** [review]

**Comment by andyross** [comment]
Correct the spelling error. As for the grammar, that's more of a style guide thing.  As a point of simple construction, a sentence of the form "X did not Y, instead X did ..." with a comma to separate the clauses, is correct.  Commas can do that?  Should they?  Meh.

I split the difference and put a semicolon in there instead.

**Pull request merged**